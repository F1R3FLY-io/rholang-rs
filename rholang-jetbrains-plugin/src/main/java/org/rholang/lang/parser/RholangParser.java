package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
