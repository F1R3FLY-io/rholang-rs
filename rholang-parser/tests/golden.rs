use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
