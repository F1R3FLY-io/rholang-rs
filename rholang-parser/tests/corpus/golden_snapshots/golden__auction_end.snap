---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: New {
                decls: [
                    NameDecl {
                        id: Id {
                            name: "anyone",
                            pos: SourcePos {
                                line: 11,
                                col: 5,
                            },
                        },
                        uri: None,
                    },
                    NameDecl {
                        id: Id {
                            name: "unused_rtn",
                            pos: SourcePos {
                                line: 11,
                                col: 13,
                            },
                        },
                        uri: None,
                    },
                ],
                proc: AnnProc {
                    proc: ForComprehension {
                        receipts: [
                            [
                                Linear {
                                    lhs: Names {
                                        names: [
                                            NameVar(
                                                Id(
                                                    Id {
                                                        name: "auction_contract",
                                                        pos: SourcePos {
                                                            line: 12,
                                                            col: 10,
                                                        },
                                                    },
                                                ),
                                            ),
                                        ],
                                        remainder: None,
                                    },
                                    rhs: Simple {
                                        name: NameVar(
                                            Id(
                                                Id {
                                                    name: "rtn",
                                                    pos: SourcePos {
                                                        line: 12,
                                                        col: 30,
                                                    },
                                                },
                                            ),
                                        ),
                                    },
                                },
                            ],
                        ],
                        proc: AnnProc {
                            proc: Send {
                                channel: Quote(
                                    AnnProc {
                                        proc: Method {
                                            receiver: AnnProc {
                                                proc: ProcVar(
                                                    Id(
                                                        Id {
                                                            name: "auction_contract",
                                                            pos: SourcePos {
                                                                line: 13,
                                                                col: 10,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 13,
                                                        col: 10,
                                                    },
                                                    end: SourcePos {
                                                        line: 13,
                                                        col: 26,
                                                    },
                                                },
                                            },
                                            name: Id {
                                                name: "get",
                                                pos: SourcePos {
                                                    line: 13,
                                                    col: 27,
                                                },
                                            },
                                            args: [
                                                AnnProc {
                                                    proc: StringLiteral(
                                                        "auction_end",
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 13,
                                                            col: 31,
                                                        },
                                                        end: SourcePos {
                                                            line: 13,
                                                            col: 44,
                                                        },
                                                    },
                                                },
                                            ],
                                        },
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 13,
                                                col: 10,
                                            },
                                            end: SourcePos {
                                                line: 13,
                                                col: 45,
                                            },
                                        },
                                    },
                                ),
                                send_type: Single,
                                inputs: [
                                    AnnProc {
                                        proc: Collection(
                                            List {
                                                elements: [
                                                    AnnProc {
                                                        proc: ProcVar(
                                                            Id(
                                                                Id {
                                                                    name: "anyone",
                                                                    pos: SourcePos {
                                                                        line: 13,
                                                                        col: 48,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 13,
                                                                col: 48,
                                                            },
                                                            end: SourcePos {
                                                                line: 13,
                                                                col: 54,
                                                            },
                                                        },
                                                    },
                                                    AnnProc {
                                                        proc: ProcVar(
                                                            Id(
                                                                Id {
                                                                    name: "unused_rtn",
                                                                    pos: SourcePos {
                                                                        line: 13,
                                                                        col: 56,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 13,
                                                                col: 56,
                                                            },
                                                            end: SourcePos {
                                                                line: 13,
                                                                col: 66,
                                                            },
                                                        },
                                                    },
                                                ],
                                                remainder: None,
                                            },
                                        ),
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 13,
                                                col: 47,
                                            },
                                            end: SourcePos {
                                                line: 13,
                                                col: 67,
                                            },
                                        },
                                    },
                                ],
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 13,
                                    col: 9,
                                },
                                end: SourcePos {
                                    line: 13,
                                    col: 68,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 12,
                            col: 5,
                        },
                        end: SourcePos {
                            line: 14,
                            col: 6,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 11,
                    col: 1,
                },
                end: SourcePos {
                    line: 15,
                    col: 2,
                },
            },
        },
    ],
)
